### PR TITLE
Add GUI Extensions Support

### PR DESCRIPTION
**Pull request by AIIX**
- Adds GUI extensions support for specific platforms, GUI Extensions provide additional platform specific functionality needed by specific supported platforms to be usable, this is a full replacement for "Platform Skills", All GUI related required functionality should remain within the GUI service itself.

- Added Support For Following GUI Extensions:
  - SmartSpeaker Extension: Provides additional functionality required that was previously provided by the skill-ovos-mycroftgui skill to connect the mycroft embedded shell front end, this extension also provides support for homescreens and homescreen management 
  - Bigscreen Extension: Provides additional functionality that was previously provided by the bigscreen platform skill for providing control over window management, Bigscreen does not provide or support homescreens
  - Generic Extension (Default): Provides generic additional functionality for generic platforms and environments that might have no additional control requirements such as Desktops, Homescreen support and Homescreen management is optionally available for this extension and has to be exclusively enabled.
  - To-Do(In Future): Add Android and Linux Mobile Extensions.

- Homescreen Manager is now independent of Service and is now loaded by the extension
- Homescreen Manager: Improvement to old style resting screens activation time and support for "mycroft.ready" signal 
- Homescreen Manager: Added show_homescreen method and handling
- Namespace: Added _get_active_extension_method() and check if extension is bigscreen during remove_namespace method for making remove namespace not clash with window close call

### CODE REVIEW

File: mycroft/configuration/mycroft.conf [hunk 1]
```diff
@@ -391,8 +391,14 @@
...
     // set this to false
-    "force_mycroft_ntp": true
+    "force_mycroft_ntp": true,
 
+    // Uncomment or add "idle_display_skill" to set initial homescreen
+    // "idle_display_skill": "skill-ovos-homescreen.openvoiceos",
+
+    // Extensions provide additional GUI platform support for specific devices
+    // Currently supported devices: SmartSpeaker, Bigscreen or Generic
+    "extension": "Generic"
```
Comment (JarbasAl) [inline mycroft/configuration/mycroft.conf:401]: can this be moved out of enclosure into the gui section of mycroft.conf?

I'm hoping we can drop "enclosure" soon and never use that word again :P

File: mycroft/gui/extensions.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            self.skill_conf = JsonStorageXDG(store_conf)
+
+        try:
+            self.bus.on("ovos.pairing.process.completed",
+                        self.start_homescreen_process)
+            self.bus.on("ovos.pairing.set.backend", self.set_backend_type)
+            self.bus.on("mycroft.gui.screen.close",
+                        self.handle_remove_namespace)
+            self.bus.on("system.reboot", self.handle_system_reboot)
+            self.bus.on("system.shutdown", self.handle_system_shutdown)
```
Comment (JarbasAl) [inline mycroft/gui/extensions.py:94]: pleae remove these, this functionality belongs in PHAL, shutdown and reboot has nothing to do with gui

File: mycroft/configuration/mycroft.conf (hunk 1 above)
Comment (AIIX) [reply mycroft/configuration/mycroft.conf:401]: I would move it to ``` "GUI": { "extension": "Generic" } ``` but "idle_display_skill" will still remain in "enclosure" if it needs to keep compact with mycroft-core, I would just like to keep them together, either i move both to "GUI" or keep them both here

File: mycroft/gui/extensions.py (hunk 2 above)
Comment (JarbasAl) [reply mycroft/gui/extensions.py:94]: the PHAL will listen to those system.xxx bus messages, PHAL does all system actions

the shell should only emit them and PHAL will call the actual commands

File: mycroft/gui/extensions.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        self.homescreen_manager = HomescreenManager(self.bus, self.gui)
+
+        self.homescreen_thread = threading.Thread(
+            target=self.homescreen_manager.run)
+        self.homescreen_thread.start()
+
+        self.device_paired = is_paired()
+        store_conf = ('skill_conf.json')
+
+        self.skill_conf = JsonStorageXDG(store_conf)
```
Comment (JarbasAl) [inline mycroft/gui/extensions.py:80]: can you explain this please? I'm not sure i understand the point of a new config file here

File: mycroft/gui/extensions.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+        except Exception as e:
+            LOG.error(f"SmartSpeaker: Init Bus Exception: {e}")
+
+    def set_backend_type(self, message):
+        backend = message.data.get("backend", "unknown")
+        if not backend == "unknown":
+            self.skill_conf["selected_backend"] = backend
+            self.skill_conf.store()
+            self.device_backend = self.skill_conf["selected_backend"]
```
Comment (JarbasAl) [inline mycroft/gui/extensions.py:106]: this doesnt seem to be used anywhere, i assume just a copy paste from old skill

is this the only usage of self.skill_conf ? i don't see anything reading from it anywhere except here

File: mycroft/gui/extensions.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+        self.name = name
+        self.bus = bus
+        self.gui = gui
+        config = Configuration.get()
+        enclosure_config = config.get("enclosure")
+        self.active_extension = enclosure_config.get("extension", "Generic")
+
+        # ToDo: Add Exclusive Support For "Desktop", "Mobile" Extensions
+        self.supported_extensions = ["SmartSpeaker", "Bigscreen", "Generic"]
```
Comment (JarbasAl) [inline mycroft/gui/extensions.py:31]: can you make this a enum at the top of the file?  to make the enum also be a regular string subclass str first

```python
class GUIExtensions(str, Enum):
     SMARTSPEAKER = "SmartSpeaker"
     BIGSCREEN = "Bigscreen"
     GENERIC = "Generic"
```


can the strings also be lower case for consistency with all plugin modules?

File: mycroft/configuration/mycroft.conf (hunk 1 above)
Comment (AIIX) [reply mycroft/configuration/mycroft.conf:401]: ok will look into it

File: mycroft/gui/extensions.py (hunk 4 above)
Comment (JarbasAl) [reply mycroft/gui/extensions.py:106]: to elaborate a bit here, users can edit mycroft.conf and change backend at any point, this new file WILL get out of sync eventually

if you really need to track this then you should read always fresh from mycroft.conf and compare the url with the official selene url

File: mycroft/gui/extensions.py (hunk 3 above)
Comment (AIIX) [reply mycroft/gui/extensions.py:80]: The config stores if the backend selected is local / remote (selene) this is due to the fact that we want to not show the homescreen at all when local backend is selected unless the device reboots on completion of the paring process. Maybe the "skill_conf" sounds like a wrong file name, but its required for knowing what backend has been selected and running through out, unless the setup skill is invoked again manually by the user and he changes the backend.

Comment (JarbasAl) [reply mycroft/gui/extensions.py:80]: i dont like this, it feels like a bug in the skill that should be solved there and not here.... its a skill specific bug not core specific

usage of override_idle should ensure homescreen doesnt show up when it should not

apart from override_idle, is the issue here the pairing check? should not report paired until the reboot, meaning the homescreen doesnt show and mycroft.ready is not sent (assuming config requires pairing for ready signal)

File: mycroft/gui/extensions.py (hunk 5 above)
Comment (JarbasAl) [reply mycroft/gui/extensions.py:31]: ignore this comment, this is good for future pluginification instead

File: mycroft/configuration/mycroft.conf (hunk 1 above)
Comment (AIIX) [reply mycroft/configuration/mycroft.conf:401]: Done

File: mycroft/gui/extensions.py (hunk 2 above)
Comment (AIIX) [reply mycroft/gui/extensions.py]: Done

File: mycroft/gui/extensions.py (hunk 3 above)
Comment (AIIX) [reply mycroft/gui/extensions.py]: Fixed and Removed

File: mycroft/gui/extensions.py (hunk 4 above)
Comment (AIIX) [reply mycroft/gui/extensions.py]: Done

File: mycroft/gui/extensions.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        get_skill_namespace = message.data.get("skill_id", "")
+        if get_skill_namespace:
+            self.bus.emit(Message("gui.clear.namespace",
+                                  {"__from": get_skill_namespace}))
+
+    def handle_system_ssh_enable(self, message):
+        ssh_enable()
+
+    def handle_system_ssh_disable(self, message):
+        ssh_disable()
```
Comment (JarbasAl) [inline mycroft/gui/extensions.py:128]: these should also be removed and part of PHAL

File: mycroft/gui/extensions.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+    def handle_remove_namespace(self, message):
+        LOG.info("Got Clear Namespace Event In Skill")
+        get_skill_namespace = message.data.get("skill_id", "")
+        if get_skill_namespace:
+            self.bus.emit(Message("gui.clear.namespace",
+                                  {"__from": get_skill_namespace}))
+
+    def handle_system_ssh_enable(self, message):
+        ssh_enable()
```
Comment (JarbasAl) [inline mycroft/gui/extensions.py:125]: these should also be removed and part of PHAL

File: mycroft/gui/extensions.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+        self.name = name
+        self.bus = bus
+        self.gui = gui
+        config = Configuration.get()
+        enclosure_config = config.get("gui")
+        self.active_extension = enclosure_config.get("extension", "Generic")
+
+        # ToDo: Add Exclusive Support For "Desktop", "Mobile" Extensions
+        self.supported_extensions = ["SmartSpeaker", "Bigscreen", "Generic"]
```
Comment (JarbasAl) [inline mycroft/gui/extensions.py:31]: please make lower case, want it to work for "generic", "GENERIC", "Generic" and "GeNeRiC"   :P

**Pull request merged**